runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/WindmillStreamPool.java

```diff
@@ -0,0 +1,177 @@
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+/**
+ * Pool of homogeneous streams to Windmill.
+ *
+ * <p>The pool holds a fixed total number of streams, and keeps each stream open for a specified
+ * time to allow for better load-balancing.
+ */
+@ThreadSafe
+public class WindmillStreamPool<StreamT extends WindmillStream> {
+
+  private final Duration streamTimeout;
+  private final Supplier<StreamT> streamSupplier;
+  private final Map<StreamT, StreamData<StreamT>> holds;
+  private final List<@Nullable StreamData<StreamT>> streamPool;
+
+  private WindmillStreamPool(
+      Duration streamTimeout,
+      Supplier<StreamT> streamSupplier,
+      List<@Nullable StreamData<StreamT>> streamPool,
+      Map<StreamT, StreamData<StreamT>> holds) {
+    this.streamPool = streamPool;
+    this.streamTimeout = streamTimeout;
+    this.streamSupplier = streamSupplier;
+    this.holds = holds;
+  }
+
+  public static <StreamT extends WindmillStream> WindmillStreamPool<StreamT> create(
+      int numStreams, Duration streamTimeout, Supplier<StreamT> streamSupplier) {
+    return new WindmillStreamPool<>(
+        streamTimeout, streamSupplier, newStreamList(numStreams), new ConcurrentHashMap<>());
+  }
+
+  @VisibleForTesting
+  static <StreamT extends WindmillStream> WindmillStreamPool<StreamT> forTesting(
+      Duration streamTimeout,
+      Supplier<StreamT> streamSupplier,
+      List<@Nullable StreamData<StreamT>> streamPool,
+      Map<StreamT, StreamData<StreamT>> holds) {
+    return new WindmillStreamPool<>(streamTimeout, streamSupplier, streamPool, holds);
+  }
+
+  /**
+   * Creates a new list of streams of the given capacity with all values initialized to null. This
+   * is because we randomly load balance across all the streams in the pool.
+   */
+  @VisibleForTesting
+  static <StreamT extends WindmillStream> List<@Nullable StreamData<StreamT>> newStreamList(
+      int numStreams) {
+    List<@Nullable StreamData<StreamT>> streamPool = new ArrayList<>(numStreams);
+    for (int i = 0; i < numStreams; i++) {
+      streamPool.add(null);
+    }
+    return streamPool;
+  }
+
+  /**
+   * Returns a stream for use that may be cached from a previous call. Each call of getStream must
+   * be matched with a call of {@link WindmillStreamPool#releaseStream(WindmillStream)}. If the
+   * stream has been cached but has timed out and drained (no longer has any holds), the stream will
+   * be closed.
+   */
+  public StreamT getStream() {
+    int index = streamPool.size() == 1 ? 0 : ThreadLocalRandom.current().nextInt(streamPool.size());
+    // We will return this stream
+    StreamT resultStream;
+    StreamT closeThisStream = null;
+    try {
+      synchronized (this) {
+        WindmillStreamPool.StreamData<StreamT> existingStreamData = streamPool.get(index);
+        // There are 3 possible states that can result from fetching the stream from the cache.
+        // 1. Stream doesn't exist create and cache a new one.
```

Comment:
nit: move the 1. and 2. comments into the if blocks like the 3. comment is, I think it will make the if-else tree more apparent.

Reply:
done.